runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -211,6 +211,372 @@ doubleMaxMinHelper(TR::Node *node, TR::CodeGenerator *cg, bool isMaxOp)
    return node->getRegister();
    }
 
+/**
+ * \brief
+ *
+ * Use vector instructions to find the index of a sub-string inside a string assuming both strings have the same element size.
+ * Each element is 1-byte for compact strings and 2-bytes for non-compressed strings.
+ *
+ * \details
+ *
+ * The vector sequence searches for the first character of the sub-string inside the source/master string. If the first character
+ * is located, it'll perform iterative vector binary compares to match the rest of the sub-string starting from the first character position.
+ *
+ *
+ * This evaluator inlines the following Java intrinsic methods:
+ *
+ * <verbatim>
+ * For Java 9 and above:
+ *
+ * StringLatin1.indexOf(s1Value, s1Length, s2Value, s2Length, fromIndex);
+ * StringUTF16.indexOf(s1Value, s1Length, s2Value, s2Length, fromIndex);
+ *
+ * For Java 8:
+ * com.ibm.jit.JITHelpers.intrinsicIndexOfStringLatin1(char[] s1Value, int s1len, char[] s2Value, int s2len, int start);
+ * com.ibm.jit.JITHelpers.intrinsicIndexOfStringUTF16(char[] s1Value, int s1len, char[] s2Value, int s2len, int start);
+ *
+ * Assumptions:
+ *
+ * -# 0 <= fromIndex < s1Length
+ * -# s1Length could be anything: positive, negative or 0.
+ * -# s2Length > 0
+ * -# s1Value and s2Value are non-null arrays and are interpreted as byte arrays.
+ * -# s1Length and s2Length are not related. i.e. s1Length could be smallers than s2Length.
+ *
+ * <\verbatim>
+ *
+ *
+ * \param node the intrinsic function call node
+ * \param cg the code generator
+ * \param elementSize element size in bytes. This is 1 for compact strings of Latin1 letters, and it's 2 for non-compact strings.
+ *
+*/
+TR::Register*
+inlineVectorizedStringIndexOf(TR::Node* node, TR::CodeGenerator* cg, uint32_t elementSize)
+   {
+   // element size is 2-byte if both s1 and s2 are decompressed strings and it is 12 otherwise.
+   // 0 for byte, 1 for halfword
+   uint32_t elementSizeMask = (elementSize == 1) ? 0 : 1;
+   bool isUTF16 = (elementSize == 2);
+
+   const int8_t vectorSize = cg->machine()->getVRFSize();
+   const bool is64Bit = TR::Compiler->target.is64Bit();
+   uintptrj_t headerSize = TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
+   TR::Compilation* comp = cg->comp();
+
+   if (comp->getOption(TR_TraceCG))
+      traceMsg(comp, "inlineVectorizedStringIndexOf element size %d\n", elementSize);
+
+   // Get call parameters
+   // s1Value and s2Value are byte arrays
+   // This evaluator function handles different indexOf() instrinsics, some of which are static calls without a receiver. Hence, the need for static call check.
+   bool isStaticCall = node->getSymbolReference()->getSymbol()
+           && node->getSymbolReference()->getSymbol()->castToMethodSymbol()
+           && node->getSymbolReference()->getSymbol()->castToMethodSymbol()->isStatic();
```

Comment:
If `node` doesn't have a method symbol associated with it I would think this whole evaluator would be wrong, hence I don't think we need to null check `node->getSymbolReference()->getSymbol()->castToMethodSymbol()` here. If you're paranoid add an assertion that it's non-null and avoid the two extra checks here.

However I suppose a better question here is why is `intrinsicIndexOfStringUTF16` not defined as `static`? Can't we just change the signature so we make sure all nodes that pass through this evaluator are `static` and we can get rid of this check and fold it away?

Reply:
Yeah, I was being paranoid about symbol checking. 

`intrinsicIndexOfStringUTF16` is not static because most methods in `com.ibm.JITHelper`  are not static. I was trying to be consistent with most JITHelper methods. 

I'll make relevant JITHelper methods static for this PR.

Reply:
The method should not be static since we use the JITHelpers instance singleton as part of the security checks around JITHelpers. All JITHelper calls in the String class should be calls to the helper instance methods which necessitates having an instance of JITHelpers and so the singleton accessor security checks will have run.

Reply:
Ah right, forgot about that. Thanks!

Reply:
Removed the symbol null checks. But the static check stays.